File: backend/src/validators/s3RequestValidator.ts
export type TemplateValidationInput = {
    requestId: string;
    parameters: Record<string, unknown>;
  };
  
  export type TemplateValidationResult = {
    valid: boolean;
    errors: string[];
    normalizedParameters?: Record<string, unknown>;
  };
  
  function normalizeBucketBaseName(value: string): string {
    return value
      .toLowerCase()
      .trim()
      .replace(/[^a-z0-9-]+/g, "-")
      .replace(/-+/g, "-")
      .replace(/^-+|-+$/g, "");
  }
  
  function isValidBucketName(value: string): boolean {
    return /^[a-z0-9][a-z0-9-]{1,61}[a-z0-9]$/.test(value);
  }
  
  function isNonEmptyString(value: unknown): value is string {
    return typeof value === "string" && value.trim() !== "";
  }
  
  export function validateAwsS3Request(
    input: TemplateValidationInput
  ): TemplateValidationResult {
    const errors: string[] = [];
    const params = { ...input.parameters };
  
    const environment =
      typeof params.environment === "string" ? params.environment : "";
  
    if (!isNonEmptyString(params.bucket_name)) {
      errors.push("Bucket name is required.");
    }
  
    if (isNonEmptyString(params.bucket_name)) {
      const normalizedBase = normalizeBucketBaseName(params.bucket_name);
  
      if (!normalizedBase) {
        errors.push("Bucket name is invalid after normalization.");
      } else {
        const shortRequestId = input.requestId.replace("req-", "").slice(0, 8);
        const finalBucketName = `${normalizedBase}-${shortRequestId}`;
  
        if (finalBucketName.length > 63) {
          errors.push(
            `Final bucket name '${finalBucketName}' exceeds the 63 character limit.`
          );
        }
  
        if (!isValidBucketName(finalBucketName)) {
          errors.push(
            `Final bucket name '${finalBucketName}' is invalid. Use lowercase letters, numbers, and hyphens only.`
          );
        }
  
        params.bucket_name = finalBucketName;
      }
    }
  
    if (!["dev", "agile", "prod"].includes(environment)) {
      errors.push("Environment must be one of: dev, agile, prod.");
    }
  
    if (!isNonEmptyString(params.tag_owner)) {
      errors.push("Tag Owner is required.");
    }
  
    if (!isNonEmptyString(params.tag_cost_center)) {
      errors.push("Tag Cost Center is required.");
    }
  
    if (!isNonEmptyString(params.tag_project)) {
      errors.push("Tag Project is required.");
    }
  
    return {
      valid: errors.length === 0,
      errors,
      normalizedParameters: errors.length === 0 ? params : undefined,
    };
  }

File: backend2/app/validators/s3_request_validator.py
def normalize_bucket_base_name(value: str) -> str:
    import re

    return re.sub(r"-+", "-", re.sub(r"[^a-z0-9-]+", "-", value.lower().strip())).strip("-")

def is_valid_bucket_name(value: str) -> bool:
    import re
    return re.fullmatch(r"[a-z0-9][a-z0-9-]{1,61}[a-z0-9]", value) is not None

def is_valid_document_name(value: str) -> bool:
    import re
    return re.fullmatch(r"[A-Za-z0-9._/-]+", value) is not None

def is_non_empty_string(value) -> bool:
    return isinstance(value, str) and value.strip() != ""

def validate_aws_s3_request(input_data: dict) -> dict:
    errors: list[str] = []
    params = dict(input_data["parameters"])

    environment = params.get("environment", "")
    environment = environment.lower().strip() if isinstance(environment, str) else ""

    website_hosting_enabled = bool(params.get("website_hosting_enabled"))

    if not is_non_empty_string(params.get("bucket_name")):
        errors.append("Bucket name is required.")

    if is_non_empty_string(params.get("bucket_name")):
        normalized_base = normalize_bucket_base_name(params["bucket_name"])

        if not normalized_base:
            errors.append("Bucket name is invalid after normalization.")
        else:
            short_request_id = input_data["requestId"].replace("req-", "")[:8]
            final_bucket_name = f"{normalized_base}-{short_request_id}"

            if len(final_bucket_name) > 63:
                errors.append(
                    f"Final bucket name '{final_bucket_name}' exceeds the 63 character limit."
                )

            if not is_valid_bucket_name(final_bucket_name):
                errors.append(
                    f"Final bucket name '{final_bucket_name}' is invalid. Use lowercase letters, numbers, and hyphens only."
                )

            params["bucket_name"] = final_bucket_name

    if environment not in ["dev", "qa", "prod", "agile"]:
        errors.append("Environment must be one of: dev, qa, agile, prod.")

    if not isinstance(params.get("encryption_enabled"), bool):
        params["encryption_enabled"] = True

    if website_hosting_enabled:
        if not is_non_empty_string(params.get("index_document")):
            errors.append("Index document is required when static website hosting is enabled.")
        elif not is_valid_document_name(params["index_document"]):
            errors.append("Index document name is invalid.")

        if is_non_empty_string(params.get("error_document")) and not is_valid_document_name(params["error_document"]):
            errors.append("Error document name is invalid.")
    else:
        params["index_document"] = ""
        params["error_document"] = ""

    if not is_non_empty_string(params.get("tag_owner")):
        errors.append("Tag Owner is required.")

    if not is_non_empty_string(params.get("tag_cost_center")):
        errors.append("Tag Cost Center is required.")

    if not is_non_empty_string(params.get("tag_project")):
        errors.append("Tag Project is required.")

    return {
        "valid": len(errors) == 0,
        "errors": errors,
        "normalizedParameters": params if len(errors) == 0 else None,
    }